### DISCUSSION TITLE
Execute a script in /sdcard location on termux via adb without physical access to remote android device

### DISCUSSION THREAD

**Discussion by tathastu871** [termux/termux-app#3249] [Q&A] [answered]
```
intent.setClassName("com.termux", "com.termux.app.RunCommandService");
intent.setAction("com.termux.RUN_COMMAND");
intent.putExtra("com.termux.RUN_COMMAND_PATH", "/data/data/com.termux/files/usr/bin/bash");
intent.putExtra("com.termux.RUN_COMMAND_ARGUMENTS", new String[]{"/storage/emulated/0/Downloads/Folder/ShellScript.sh"});
intent.putExtra("com.termux.RUN_COMMAND_WORKDIR", "/data/data/com.termux/files/home");
intent.putExtra("com.termux.RUN_COMMAND_BACKGROUND", false);
intent.putExtra("com.termux.RUN_COMMAND_SESSION_ACTION", "0");
startService(intent);
```
can someone provide adb am start intent command for this.
after installint fresh termux i want to execute certain command in termux without opening termux directly from adb pc

on remote system i have termux insyalled freshly and it has setup.sh in /sdcard
device has adb enabled
i want to
 cp /sdcard/setup.sh $HOME/ and bash $HOME/setup.sh
from adb shell without actually opening and typing in termux as it is on remote device i dont have physical  acces to 
(its a fresh install it doesnt have sshd installed or configured so its not an option)
on non root device with non debug termux build

i have seen mixplorer app to run termux commands from app itself 
can i do same with adb

**Comment by agnostic-apollo** [comment, marked as answer]
The `adb` `shell` user does not request `RUN_COMMAND` permission, so it cannot send `RUN_COMMAND` intents. You could use `run-as` command over adb if using github debug builds (not F-Droid) or if you have rooted your device, then can send intent with `su` inside the `adb shell`.

**Reply by tathastu871** [reply]
lets say i install debug build on non root device
can you give me adb run-as command to fulfill my need

also can debug build be autoupdated from fdroid. or i have to update it manually

**Reply by agnostic-apollo** [reply]
Note that it is not safe to run scripts from sd card since other malicious apps could modify the scripts, allowing them to run code in termux or root context. Use `run-as` directly.

```shell
termux_package_name="com.termux"
termux_activity_name=".app.TermuxActivity"

# The 4 `TERMUX__` variables will automatically be exported in v0.119.0+ and will exist in termux.env file
termux_init_command="/system/bin/sh -c 'TERMUX__ROOTFS=\"/data/data/$termux_package_name/files\""
termux_init_command+="$(cat <<'EOF'

TERMUX__HOME="$TERMUX__ROOTFS/home"
TERMUX__PREFIX="$TERMUX__ROOTFS/usr"
TERMUX__DATA_HOME="$TERMUX__HOME/.termux"

# Wait till bootstrap is installed
i=0;
while true; do
    sleep 1
    if [ -d "$TERMUX__PREFIX" ] && [ -d "$TERMUX__HOME" ]; then
        echo "Termux bootstrap setup complete"
        break
    fi
    if [ $i -gt 30 ]; then
        echo "Termux bootstrap not setup even after 30s" 1>&2
        exit 1
    fi
    i=$((i+1))
done

# Source termux.env file
. "$TERMUX__PREFIX/etc/termux/termux.env" || exit 1

# Create TERMUX__DATA_HOME if it does not already exist
echo "creating termux data home"
mkdir -p "$TERMUX__DATA_HOME" || exit 1
chmod 700 "$TERMUX__DATA_HOME" || exit 1

# Enable allow-external-apps property
# Or run something else
echo "Enabling allow-external-apps"
prop_value="true"; prop_key="allow-external-apps"; 
termux_properties_file="$TERMUX__DATA_HOME/termux.properties";
if ! grep -E '^'"$prop_key"'=.*' $termux_properties_file &>/dev/null; then
    [ -s "$termux_properties_file" ] && [ ! -z "$(tail -c 1 "$termux_properties_file")" ] && newline=$'\n' || newline="";
    echo "$newline$prop_key=$prop_value" >> "$termux_properties_file" || exit 1
else
    sed -i'' -E 's/^'"$prop_key"'=.*/'"$prop_key=$prop_value"'/' $termux_properties_file || exit 1
fi

# Update termux-tools package
apt-get update -y || exit 1
apt-get install -y -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" termux-tools || exit 1
'
EOF
)"

adb shell am force-stop "$termux_package_name" &&
adb shell am start --user 0 "$termux_package_name/$termux_activity_name" &&
adb shell run-as "$termux_package_name" --user 0 "$termux_init_command"
```

**Reply by tathastu871** [reply]
will above script work in cygwin. 
but how to execute setup.sh i have. it just has some apt commands and cp and tar and netcat commands
cat i just 
cat heredoc all those commands as shown in above script.

eg

```
termux_package_name="com.termux"
termux_activity_name=".app.TermuxActivity"

termux_init_command="/system/bin/sh -c 'TERMUX__ROOTFS=\"/data/data/$termux_package_name/files\""
termux_init_command+="$(cat <<'EOF'
apt update -y
tar -zxvf /sdcard/file.tar.gz
)"
```

**Reply by agnostic-apollo** [reply]
cygwin or WSL bash shell should work. 

The script should likely work. Use `apt-get update -y || exit 1` instead  of `apt` since `apt` should not be run in scripts.

https://github.com/termux/termux-app/discussions/2652#discussioncomment-2339613

Some commands may require `LD_PRELOAD`, so will require starting a new termux `bash` shell to run the commands.

**Comment by tathastu871** [comment]
can dev give a common adb shel pm or adb shell appops 
command to give all necessary permissions that termux.apk require
including 
draw over other apps.

i tried to extract permission from manifest and supply them to  xargs -n1  adb shell pm grant
but java exception errors occur 
just want to install 

common script to
1)install termux.apk over adb
2)grant all permission including exclusive permission that can only be granted by adb like appops
please. thanks in advance

**Reply by agnostic-apollo** [reply]
For draw over apps, run `adb shell "appops set --user 0 com.termux SYSTEM_ALERT_WINDOW allow"`

For storage, run `adb shell "pm grant com.termux android.permission.READ_EXTERNAL_STORAGE; pm grant com.termux android.permission.WRITE_EXTERNAL_STORAGE"`

For disabling battery optimizations, run `adb shell "dumpsys deviceidle whitelist +com.termux"`

**Reply by tathastu871** [reply]
> For draw over apps, run `adb shell "appops set --user 0 com.termux SYSTEM_ALERT_WINDOW allow"`
> 
> For storage, run `adb shell "pm grant com.termux android.permission.READ_EXTERNAL_STORAGE; pm grant com.termux android.permission.WRITE_EXTERNAL_STORAGE"`
> 
> For disabling battery optimizations, run `adb shell "dumpsys deviceidle whitelist +com.termux"`


are above only adb grantable permission or are there many other in manifest.xml



```
termux_package_name="com.termux"
termux_activity_name=".app.TermuxActivity"

termux_init_command="/system/bin/sh -c 'TERMUX__ROOTFS=\"/data/data/$termux_package_name/files\""
termux_init_command+="$(cat <<'EOF'
apt update -y
tar -zxvf /sdcard/file.tar.gz
)"
```
is above script correct as i understand. will it work in cygwin a it is

**Reply by agnostic-apollo** [reply]
Some are automatically granted by android, some require adb or root like `READ_LOGS`, `WRITE_SECURE_SETTINGS`, `DUMP`, `PACKAGE_USAGE_STATS`.

**Comment by tbultel** [comment]
Thanks for the script that uses "run-as"
However, it fails because it cannot find termux.env, that seems to be nowhere on  the filesystem.
I am using termux-app_v0.118.0+github-debug_universal.apk
Waiting for someone's answer about this, I am digging in the sources to find out why.

**Reply by agnostic-apollo** [reply]
It's not in `0.118.0`, use latest github actions builds.

https://github.com/termux/termux-app#github

https://github.com/termux/termux-app/commit/03e1d14e1e65e054eeeb0ad672a46c2c17254d3e
